### PR TITLE
feat(compiler-cli): run JIT transform on classes with jit: true opt-out

### PR DESCRIPTION
**Pull request by devversion**
Currently when compiling code with the Angular compiler, all classes
with Angular decorators are compiled with AOT. This includes type
checking, scope collection etc.

This may not be desirable for all components, e.g. dynamic components,
or test components w/ `TestBed.configureTestingModule` (if compiled with ngtsc).

Those components can opt out of AOT on a per component-basis via `jit:
true`. 
This is helpful as it allows incremental migrations/refactorings to AOT.
Whether we want to keep this capability long-term is something to be discussed separately.

For now though, we should fix that components compiled with `jit: true`
actually work as expected. Currently this **not the case** as soon as
the new initializer APIs are used— as those do no longer declare class
metadata with decorators.

This commit runs the JIT transform on JIT-opted classes.

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/annotations/common/src/jit_declaration_registry.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ */
+
+import {ClassDeclaration} from '../../../reflection';
+
+/**
+ * Registry that keeps track of Angular declarations that are explicitly
+ * marked for JIT compilation and are skipping compilation by trait handlers.
+ */
+export class JitDeclarationRegistry {
+  jitDeclarations: ClassDeclaration[] = [];
```
Comment (crisbeto) [inline packages/compiler-cli/src/ngtsc/annotations/common/src/jit_declaration_registry.ts:16]: nit: should this be a `Set` or we don't run the risk of duplicates?

File: packages/compiler-cli/src/ngtsc/annotations/directive/src/handler.ts [hunk 2]
```diff
@@ -98,6 +99,8 @@ const LIFECYCLE_HOOKS = new Set([
   'ngAfterContentChecked',
 ]);
 
+export const JIT_ANALYSIS_MARKER = Symbol();
```
Comment (crisbeto) [inline packages/compiler-cli/src/ngtsc/annotations/directive/src/handler.ts:102]: This doesn't seem to be used anywhere.

**Review by crisbeto — APPROVED** [review]

**Review by pkozlowski-opensource — APPROVED** [review]

File: packages/compiler-cli/src/ngtsc/annotations/common/src/jit_declaration_registry.ts (hunk 1 above)
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/annotations/common/src/jit_declaration_registry.ts:16]: Done.

File: packages/compiler-cli/src/ngtsc/annotations/directive/src/handler.ts (hunk 2 above)
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/annotations/directive/src/handler.ts:102]: Good catch

**Pull request closed without merging**